pages/index.js

```diff
@@ -1,8 +1,51 @@
-import React from 'react';
+import React, { useState } from 'react';
 import Link from 'next/link';
 import { getIcon } from '../icons/utils';
 import { getAllPosts } from '../lib/api';
 
+const lcs = (stringPar, patternPar) => {
+  const string = stringPar.toLowerCase();
+  const pattern = patternPar.toLowerCase();
+
+  const m = string.length;
+  const n = pattern.length;
+
+  const dp = [];
+  let result = 0;
+  for (let i = 0; i < m + 1; i++) dp.push(Array(n + 1));
+
+  for (let i = 0; i < m + 1; i++) {
+    for (let j = 0; j < n + 1; j++) {
+      if (i === 0 || j === 0) dp[i][j] = 0;
+      else if (string[i - 1] === pattern[j - 1]) {
+        dp[i][j] = dp[i - 1][j - 1] + 1;
+        result = Math.max(result, dp[i][j]);
+      } else dp[i][j] = 0;
+    }
+  }
+
+  return result;
+};
+
+const useFilter = posts => {
+  const [filteredPosts, setFilteredPosts] = useState(posts);
+
+  const onKeywordsChange = e => {
+    const keywords = e.target.value;
+    setFilteredPosts(
+      posts
+        .map(post => {
+          const postData = [post.title, post.description, post.content].join(' ');
+          post.matchedLength = lcs(postData, keywords);
+          return post;
+        })
+        .sort((post1, post2) => (post1.matchedLength > post2.matchedLength ? '-1' : '1'))
```

Comment:
Line 39 breaks ESLint's no-param-reassign rule. Can you think of a better way to do this ?

Reply:
yeah, you shouldn't ever try to modify params. `map` is a *transformation* of each element of the list. In this case, transforming means that you compute a new list. You are not supposed to modify the list you are mapping over, that's a source of bugs...state should be immutable when possible (cause we're in javascript :grin:).

```suggestion
        .map(post => {
          const postData = [post.title, post.description, post.content].join(' ');
          
          return {
              ...post,
              matchedLength: lcs(postData, keywords)
          };
        })
        .sort((post1, post2) => (post1.matchedLength > post2.matchedLength ? '-1' : '1'))
```